src/index.js

```diff
@@ -59,10 +74,23 @@ class CID {
    * new CID(<bs58 encoded multihash>)
    * new CID(<cid>)
    */
-  constructor (version, codec, multihash, multibaseName = 'base58btc') {
-    if (module.exports.isCID(version)) {
-      // version is an exising CID instance
-      const cid = version
+  /*::
+  +version:Version
+  +codec:Codec
+  +multihash:Multihash
+  +multibaseName:MultibaseName
+  string:?string
+  _buffer:?Buffer
+  static isCID:(mixed) => boolean
+  */
+  constructor (
+    version/*: CID<a> | BaseEncodedString | Buffer | Version */,
+    codec/*: ?Codec */ = undefined,
+    multihash/*: ?Multihash */ = undefined,
+    multibaseName/*: MultibaseName */ = 'base58btc'
+  ) {
```

Comment:
This API is a nightmare from type-checker standpoint. I think size & code flow in the construction also reflects the fact that it's pretty hairy. I'm proposing https://github.com/multiformats/js-cid/issues/83 to clear this up.